File: LeetCode/501二叉搜索树中的众数.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int *ans;
int ansSize, base, cnt, max_cnt;

void update(int x) {
    if (x == base) cnt++;
    else {
        base = x;
        cnt = 1;
    }
    if (cnt == max_cnt) {
        ans[ansSize++] = base;
    }
    if (max_cnt < cnt) {
        max_cnt = cnt;
        ansSize = 0;
        ans[ansSize++] = base;
    }
}

void order(struct TreeNode *root) {
    if (root == NULL) return ;
    order(root->left);
    update(root->val);
    order(root->right);
    return ;
}

int* findMode(struct TreeNode* root, int* returnSize){
    base = cnt = max_cnt = ansSize = 0;
    ans = (int *)malloc(sizeof(int) * 4001);
    order(root);
    *returnSize = ansSize;
    return ans;
}


/**
 * C++
 */
class Solution {
public:
    void preOrder(TreeNode* root, map<int, int> &map) {
        if (root == nullptr) return ;
        map[root->val]++;
        preOrder(root->left, map);
        preOrder(root->right, map);
        return ;
    }

    static bool cmp(pair<int, int>& a, pair<int, int>& b) {
        return a.second > b.second;
    }

    vector<int> findMode(TreeNode* root) {
        map<int, int> map;
        vector<int> ans;
        if (root == nullptr) return ans;
        preOrder(root, map);
        vector<pair<int, int>> vec(map.begin(), map.end());
        sort(vec.begin(), vec.end(), cmp);
        ans.push_back(vec[0].first);
        for (int i = 1; i < vec.size(); i++) {
            if (vec[i].second == vec[0].second) ans.push_back(vec[i].first);
            else break;
        }
        return ans;
    }
};

File: LeetCode/501二叉搜索树中的众数.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findMode(root *TreeNode) []int {
	sta := []*TreeNode{}
	cur := root
	var pre *TreeNode
	cnt, maxCnt := 0, 0
	ans := []int{}
	for cur != nil || len(sta) > 0 {
		if cur != nil {
			sta = append(sta, cur)
			cur = cur.Left
		} else {
			cur = sta[len(sta)-1]
			sta = sta[:len(sta)-1]
			if pre == nil { //第一个节点
				cnt = 1
			} else if pre.Val == cur.Val { //与前一个节点相同
				cnt++
			} else { //与前一个节点不同
				cnt = 1
			}
			if cnt == maxCnt {
				ans = append(ans, cur.Val)
			}
			if cnt > maxCnt {
				maxCnt = cnt
				ans = []int{}
				ans = append(ans, cur.Val)
			}
			pre = cur
			cur = cur.Right
		}
	}
	return ans
}

/*
	整体框架是使用栈模拟二叉树的中序遍历，也就是迭代法
	加上两个控制条件，一个是统计频率，一个统计最大频率
	当cnt == maxCnt 时候，表示访问到了一个值，但是也有可能maxCnt还会变大
	因此如果cnt > maxCnt需要更新maxCnt，然后清空切片，再把数加进去
*/